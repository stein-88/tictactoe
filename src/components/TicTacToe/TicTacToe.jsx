import React from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { selectCell, resetBoard, changePage } from '@actions/moves'
import StartGame from './StartGame'
import EndGame from './EndGame'
import Gaming from './Gaming'
import s from './TicTacToe.scss'

const selectBoard = (state) => state.board
const selectGame = (state) => state.game
const selectPage = (state) => state.page

const TicTacToe = () => {
	const dispatch = useDispatch()
	const board = useSelector(selectBoard)
	const game = useSelector(selectGame)
	const selectedPage = useSelector(selectPage)

	const changeToPage = () => dispatch(changePage())
	const resetGame = () => dispatch(resetBoard())

	const handleDispatch = (row, col) => {
		if (!game || !game.currentPlayer) return null
		return dispatch(selectCell(game.currentPlayer, row, col))
	}
	return (
		<section className={`w-100 ${s.bgNewlikeblue}`}>
			<div className="container">
				<div className="row">
					<div className="col-lg-8 col-12 m-auto">
						{selectedPage === 1
							&& <StartGame changeToPage={changeToPage} />}
						{selectedPage === 2
							&& (
								<Gaming
									resetGame={resetGame}
									handleDispatch={handleDispatch}
									currentPlayer={game && game.currentPlayer}
									board={board}
								/>
							)}
						{selectedPage === 3
							&& <EndGame restartGame={resetGame} winner={game.winner} />}
					</div>
				</div>
			</div>
		</section>
	)
}

export default TicTacToe
