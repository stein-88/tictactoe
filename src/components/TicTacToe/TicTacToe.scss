.bgNewlikeblue {
  background-color: #cfd0f9;
  padding: 8rem 0;
}
