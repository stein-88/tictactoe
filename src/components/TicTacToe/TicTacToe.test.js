import React from 'react'
import renderer from 'react-test-renderer'
import TicTacToe from './TicTacToe'
import * as reactRedux from 'react-redux'
import configureStore from '../../store'
import schema from './StartGame/data'
const { players } = schema
const { Provider } = reactRedux
reactRedux.useDispatch = () => jest.fn()

jest.mock('uuid', () => ({ v4: () => `uniquekey${Math.random(1000) * 100}` }))

describe('Test on TicTacToe Component', () => {
	test('Should render without props and start on StartGame page', () => {
		const compTicTacToe = renderer.create(
			<Provider store={configureStore()}>
				<TicTacToe />
			</Provider>
		)
		const jsonTicTacToe = compTicTacToe.toJSON()
		const Result = compTicTacToe.root.findAllByProps({
			['data-test']: 'players'
		})
		const Results = compTicTacToe.root.findAllByProps({
			className: 'btnstyle'
		})[0]
		Results.props.onClick()
		expect(Result).toHaveLength(players.length)
		expect(jsonTicTacToe).toMatchSnapshot()
	})
	test('Sould render gamepage inside of tictoctoe', () => {
		const store = configureStore()
		store.getState().page = 2
		const compTicTacToe = renderer.create(
			<Provider store={store}>
				<TicTacToe />
			</Provider>
		)
		const jsonTicTacToe = compTicTacToe.toJSON()
		const Result = compTicTacToe.root.findAllByProps({
			['data-testrefer']: ''
		})[0]
		Result.props.onClick()
		expect(jsonTicTacToe).toMatchSnapshot()
	})
	test('handleDispatch should return null', () => {
		const store = configureStore()
		store.getState().page = 2
		const compTicTacToe = renderer.create(
			<Provider store={store}>
				<TicTacToe />
			</Provider>
		)
		const jsonTicTacToe = compTicTacToe.toJSON()
		const Result = compTicTacToe.root.findAllByProps({
			['data-testrefer']: ''
		})[0]
		store.getState().game.currentPlayer = null
		Result.props.onClick()
		expect(jsonTicTacToe).toMatchSnapshot()
	})
	test('Should render endGame inside tictactoe', () => {
		const store = configureStore()
		store.getState().page = 3
		store.getState().game = {
			winner: 'Tie'
		}
		const compTicTacToe = renderer.create(
			<Provider store={store}>
				<TicTacToe />
			</Provider>
		)
		const jsonTicTacToe = compTicTacToe.toJSON()
		const Result = compTicTacToe.root.findAllByProps({
			className: 'btnstyle'
		})[0]
		Result.props.onClick()
		expect(jsonTicTacToe).toMatchSnapshot()
	})
})