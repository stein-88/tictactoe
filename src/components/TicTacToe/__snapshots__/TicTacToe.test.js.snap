// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on TicTacToe Component Should render endGame inside tictactoe 1`] = `
<section
  className="w-100 bgNewlikeblue"
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-lg-8 col-12 m-auto"
      >
        <div
          className="row"
        >
          <div
            className="col-lg-9 col-12 m-auto"
          >
            <section
              className="w-100 text-center"
            >
              <p
                className="h1"
              >
                Tie
              </p>
              <section
                className="w-100 px-5"
              >
                <img
                  alt=""
                  className="w-100 img-fluid rounded py-3 px-5"
                  data-test="placarimg"
                  src="test-file-stub"
                />
              </section>
              <span
                aria-hidden="true"
                className="btnstyle"
                onClick={[Function]}
              >
                Play again
              </span>
            </section>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;

exports[`Test on TicTacToe Component Should render without props and start on StartGame page 1`] = `
<section
  className="w-100 bgNewlikeblue"
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-lg-8 col-12 m-auto"
      >
        <div
          className="row"
        >
          <div
            className="col-lg-6 col-12"
          >
            <p
              className="h1 pb-5"
            >
              Tic Tac Toe
            </p>
            <p
              className="h3 py-3"
              data-test="players"
            >
              Player One: 
              <strong
                className="text-danger"
              >
                X
              </strong>
            </p>
            <p
              className="h3 "
              data-test="players"
            >
              Player Two: 
              <strong
                className="text-danger"
              >
                O
              </strong>
            </p>
          </div>
          <div
            className="col-lg-6 col-12"
          >
            <section
              className="w-100 text-center"
            >
              <section
                className="w-100 px-5 mb-3"
              >
                <img
                  alt="Tic Tac Toe"
                  className="w-100 img-fluid rounded"
                  src="test-file-stub"
                />
              </section>
              <span
                aria-hidden="true"
                className="btnstyle"
                onClick={[Function]}
              >
                Start Game
              </span>
            </section>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;

exports[`Test on TicTacToe Component Sould render gamepage inside of tictoctoe 1`] = `
<section
  className="w-100 bgNewlikeblue"
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-lg-8 col-12 m-auto"
      >
        <section
          className="w-100"
        >
          <section
            className="w-100 text-right pb-3"
          >
            <span
              aria-hidden="true"
              className="btnstyle"
              onClick={[Function]}
            >
              Reset
            </span>
          </section>
          <section
            className="w-100"
          >
            <div
              className="col-lg-9 col-12 m-auto"
            >
              <div
                className="row m-0"
              >
                <div
                  aria-hidden="true"
                  className="col clickText  tictok box0col0"
                  data-testrefer=""
                  onClick={[Function]}
                >
                  
                </div>
                <div
                  aria-hidden="true"
                  className="col clickText  tictok box0col1"
                  data-testrefer=""
                  onClick={[Function]}
                >
                  
                </div>
                <div
                  aria-hidden="true"
                  className="col clickText  tictok box0col2"
                  data-testrefer=""
                  onClick={[Function]}
                >
                  
                </div>
              </div>
            </div>
            <div
              className="col-lg-9 col-12 m-auto"
            >
              <div
                className="row m-0"
              >
                <div
                  aria-hidden="true"
                  className="col clickText  tictok box1col0"
                  data-testrefer=""
                  onClick={[Function]}
                >
                  
                </div>
                <div
                  aria-hidden="true"
                  className="col clickText  tictok box1col1"
                  data-testrefer=""
                  onClick={[Function]}
                >
                  
                </div>
                <div
                  aria-hidden="true"
                  className="col clickText  tictok box1col2"
                  data-testrefer=""
                  onClick={[Function]}
                >
                  
                </div>
              </div>
            </div>
            <div
              className="col-lg-9 col-12 m-auto"
            >
              <div
                className="row m-0"
              >
                <div
                  aria-hidden="true"
                  className="col clickText  tictok box2col0"
                  data-testrefer=""
                  onClick={[Function]}
                >
                  
                </div>
                <div
                  aria-hidden="true"
                  className="col clickText  tictok box2col1"
                  data-testrefer=""
                  onClick={[Function]}
                >
                  
                </div>
                <div
                  aria-hidden="true"
                  className="col clickText  tictok box2col2"
                  data-testrefer=""
                  onClick={[Function]}
                >
                  
                </div>
              </div>
            </div>
          </section>
          <section
            className="w-100 text-right pt-3"
          >
            <p
              className="h3"
            >
              Turn of the Player  X
            </p>
          </section>
        </section>
      </div>
    </div>
  </div>
</section>
`;

exports[`Test on TicTacToe Component handleDispatch should return null 1`] = `
<section
  className="w-100 bgNewlikeblue"
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="col-lg-8 col-12 m-auto"
      >
        <section
          className="w-100"
        >
          <section
            className="w-100 text-right pb-3"
          >
            <span
              aria-hidden="true"
              className="btnstyle"
              onClick={[Function]}
            >
              Reset
            </span>
          </section>
          <section
            className="w-100"
          >
            <div
              className="col-lg-9 col-12 m-auto"
            >
              <div
                className="row m-0"
              >
                <div
                  aria-hidden="true"
                  className="col clickText  tictok box0col0"
                  data-testrefer=""
                  onClick={[Function]}
                >
                  
                </div>
                <div
                  aria-hidden="true"
                  className="col clickText  tictok box0col1"
                  data-testrefer=""
                  onClick={[Function]}
                >
                  
                </div>
                <div
                  aria-hidden="true"
                  className="col clickText  tictok box0col2"
                  data-testrefer=""
                  onClick={[Function]}
                >
                  
                </div>
              </div>
            </div>
            <div
              className="col-lg-9 col-12 m-auto"
            >
              <div
                className="row m-0"
              >
                <div
                  aria-hidden="true"
                  className="col clickText  tictok box1col0"
                  data-testrefer=""
                  onClick={[Function]}
                >
                  
                </div>
                <div
                  aria-hidden="true"
                  className="col clickText  tictok box1col1"
                  data-testrefer=""
                  onClick={[Function]}
                >
                  
                </div>
                <div
                  aria-hidden="true"
                  className="col clickText  tictok box1col2"
                  data-testrefer=""
                  onClick={[Function]}
                >
                  
                </div>
              </div>
            </div>
            <div
              className="col-lg-9 col-12 m-auto"
            >
              <div
                className="row m-0"
              >
                <div
                  aria-hidden="true"
                  className="col clickText  tictok box2col0"
                  data-testrefer=""
                  onClick={[Function]}
                >
                  
                </div>
                <div
                  aria-hidden="true"
                  className="col clickText  tictok box2col1"
                  data-testrefer=""
                  onClick={[Function]}
                >
                  
                </div>
                <div
                  aria-hidden="true"
                  className="col clickText  tictok box2col2"
                  data-testrefer=""
                  onClick={[Function]}
                >
                  
                </div>
              </div>
            </div>
          </section>
          <section
            className="w-100 text-right pt-3"
          >
            <p
              className="h3"
            >
              Turn of the Player  X
            </p>
          </section>
        </section>
      </div>
    </div>
  </div>
</section>
`;
